x/feedist/handler.go

```diff
@@ -0,0 +1,300 @@
+	}
+
+	rev.Recipients = msg.Recipients
+	obj := orm.NewSimpleObj(msg.RevenueID, rev)
+	if err := h.bucket.Save(db, obj); err != nil {
+		return res, errors.Wrap(err, "cannot save")
+	}
+	return res, nil
+}
+
+func (h *updateRevenueHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*UpdateRevenueMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get message")
+	}
+	msg, ok := rmsg.(*UpdateRevenueMsg)
+	if !ok {
+		return nil, errors.InvalidMsgErr.New("unknown transaction type")
+	}
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+	return msg, nil
+}
+
+// distribute split the funds stored under the revenue address and distribute
+// them according to recipients proportions. When successful, revenue account
+// has no funds left after this call.
+//
+// It might be that not all funds can be distributed equally. Because of that a
+// small leftover can remain on the revenue account after this operation.
+func distribute(db weave.KVStore, ctrl CashController, source weave.Address, recipients []*Recipient) error {
+	var chunks int64
+	for _, r := range recipients {
+		chunks += int64(r.Weight)
+	}
+
+	// Find the greatest common division for all weights. This is needed to
+	// avoid leaving big fund leftovers on the source account when
+	// distributing between many recipients. Or when there is only one
+	// recipient with a high weight value.
+	var weights []int32
+	for _, r := range recipients {
+		weights = append(weights, r.Weight)
+	}
+	div := findGcd(weights...)
+
+	chunks = chunks / int64(div)
+
+	balance, err := ctrl.Balance(db, source)
+	switch {
+	case err == nil:
+		// All good.
+	case errors.Is(errors.NotFoundErr, err):
+		// Account does not exist, so there is are no funds to split.
+		return nil
+	default:
+		return errors.Wrap(err, "cannot acquire revenue account balance")
+	}
+
+	// TODO normalize balance. There is no functionality that allows to
+	// normalize x.Coins right now (14 Feb 2019).
+
+	// For each currency, distribute the coins equally to the weight of
+	// each recipient. This can leave small amount of coins on the original
+	// account.
+	for _, c := range balance {
+		// Ignore those coins that have a negative value. This
+		// functionality is supposed to be distributing value from
+		// revenue account, not collect it. Otherwise this could be
+		// used to charge the recipients instead of paying them.
+		if !c.IsPositive() {
+			continue
+		}
+
+		// Rest of the division can be ignored, because we transfer
+		// funds to each recipients separately. Any leftover will be
+		// left on the recipients account.
+		one, _ := c.Divide(chunks)
+
+		for _, r := range recipients {
+			amount := one.Multiply(int64(r.Weight / div))
+			// Chunk is too small to be distributed.
+			if amount.IsZero() {
+				continue
+			}
+			if err := ctrl.MoveCoins(db, source, r.Address, amount); err != nil {
+				return errors.Wrap(err, "cannot move coins")
+			}
```

Comment:
req: We can have a sanity check that makes sure we are not loosing/adding any money with this: "old balance - amounts sent  == rest". See #322 
Ideally we would move this sanity check to the cash package level to protect all handlers but this is likely off scope of this ticket. Just mentioning it here for a quick 👍 👎

Reply:
I believe this is tested in the coins package.
And it ensures it never goes beyond 0, so if you try to distribute more funds than you have, this will stop you

Reply:
I could not find the check in BaseController.MoveCoins but unmoved coins will remain on the feedist account anyway nor we could add new coins by rounding issues here.